### PR TITLE
Issue #4607: Added moduleId to violation messages

### PR DESCRIPTION
**Pull request by Luolc**
#4607 

In XMLLogger, there would be an `id="..."` if the module id is set. (and no `source="..."`)

Log message would be `[SEVERITY LEVEL] filePath:lineNo:columnNo: message. [ModuleId]`, if module id is set.

I was stuck before for I didn't realize that I need to modify `compareTo` in LocalizedMessage.

### CODE REVIEW

**Comment by romani** [comment]
CIs should be green

File: src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java [hunk 1]
```diff
@@ -415,7 +416,11 @@ public int compareTo(LocalizedMessage other) {
 
         if (lineNo == other.lineNo) {
             if (columnNo == other.columnNo) {
-                result = getMessage().compareTo(other.getMessage());
+                result = Comparator.nullsFirst(Comparator.<String>naturalOrder())
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java]: @romani 
> Classes which are missed in Cobertura coverage report:
com/puppycrawl/tools/checkstyle/api/LocalizedMessage

It looks like either `Comparator.nullsFirst` or `Comparator.naturalOrder` forces cobertura to ignore the entire class like we have seen with `streams`.

@Luolc Please rewrite this class not using those 2 methods and figure out which is causing the problem.

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java]: @Luolc , please add such classes to known list - https://github.com/checkstyle/checkstyle/blob/master/config/import-control.xml#L19 to disallow usage of them till we are on cobertura. Such update could be separate PR, to avoid mixing too much in one PR.

**Comment by Luolc** [comment]
@romani I just added a UT as rnveach said at https://github.com/checkstyle/checkstyle/pull/4899#pullrequestreview-54487344

I cannot run verify currently due to #4842. CI failed and I would have a check after I get home to use my Mac

**Comment by romani** [comment]
> I cannot run verify currently due

If problem is only at `MainTest.testCreateListenerWithLocationIllegalStateException`, you can exclude it execution `-Dtest='*,!MainTest'` , it will fail on missed coverage, but you can recheck coverage manually.

**Comment by rnveach** [comment]
> I cannot run verify currently due to #4842.

@Luolc I get around this issue currently by running `verify` in pieces similar to like travis does. Look at `travis.yml` to see the commands separated out.

**Comment by Luolc** [comment]
@romani @rnveach Thanks for the help.

I updated the IDEA inspection config. There would be two more classes in `CheckerTest`.
UT for https://github.com/checkstyle/checkstyle/pull/4899#pullrequestreview-54487344 is placed in `CheckerTest`. I am not very sure if this is the proper place.

File: config/intellij-idea-inspections.xml [hunk 2]
```diff
@@ -332,7 +332,7 @@
     </inspection_tool>
     <inspection_tool class="ClassWithOnlyPrivateConstructors" enabled="true" level="WARNING" enabled_by_default="true" />
     <inspection_tool class="ClassWithTooManyDependencies" enabled="true" level="ERROR" enabled_by_default="true">
-        <option name="limit" value="34" />
+        <option name="limit" value="36" />
```
Comment (rnveach) [inline config/intellij-idea-inspections.xml]: @romani We just upped this number because we couldn't find a way to suppress it for a specific file, and here we are upping it again.
Should we just turn it off and report problem to IntelliJ?

File: src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java [hunk 3]
```diff
@@ -129,9 +129,16 @@ public void addError(AuditEvent event) {
...
-            writer.println(" source=\""
-                + encode(event.getSourceName())
-                + "\"/>");
+            if (event.getModuleId() == null) {
+                writer.println(" source=\""
+                        + encode(event.getSourceName())
+                        + "\"/>");
+            }
+            else {
+                writer.println(" id=\""
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java]: Why is this surrounded by an `if` if the attributes for both are different. We can have both in XML file if we wish.

Since we are completely overriding the `source` in the CLI output to only show the ID, we should do the exact same in the XML.
Make this attribute name `source` and not `id`.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java [hunk 4]
```diff
@@ -411,16 +411,30 @@ public static void setLocale(Locale locale) {
...
     @Override
     public int compareTo(LocalizedMessage other) {
-        int result = Integer.compare(lineNo, other.lineNo);
+        final int result;
 
         if (lineNo == other.lineNo) {
             if (columnNo == other.columnNo) {
-                result = getMessage().compareTo(other.getMessage());
+                if (Objects.equals(moduleId, other.moduleId)) {
+                    result = getMessage().compareTo(other.getMessage());
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java]: @romani I don't see us comparing non-id sources of message.
If 2 different modules print the exact same message on the same line/column, we will have the same issue as before.
Should this be a new issue where we add module name if `id` is not specified to the compare?

File: src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java [hunk 5]
```diff
@@ -104,6 +104,23 @@ public void testFormatModuleNameDoesNotContainCheckSuffix() {
     }
 
     @Test
+    public void testFormatModuleWithModuleId() {
+        final AuditEvent mock = PowerMockito.mock(AuditEvent.class);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java]: Why is this a mock? Why can't we instantiate `AuditEvent` and `LocalizedMessage`?
Mocks should be a last resort when deep code can be done normally.

You were able to get by without a mock in `XMLLoggerTest`.

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java [hunk 6]
```diff
@@ -1010,6 +1020,63 @@ public void testXmlLoggerClosesItStreams() throws Exception {
...
+        final DefaultConfiguration root = new DefaultConfiguration("root");
+        root.addChild(moduleConfig1);
+        root.addChild(moduleConfig2);
+        final Checker checker = new Checker();
+        checker.setModuleClassLoader(Thread.currentThread().getContextClassLoader());
+        checker.configure(root);
+        // BriefUtLogger does not print the module name or id postfix,
+        // so we need to set logger manually
+        final ByteArrayOutputStream out =
+                (ByteArrayOutputStream) Whitebox.getInternalState(this, "stream");
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java]: From here until the end, why can't we call `verify(Checker checker, String fileName, String... expected)`?
Duplicating all this code isn't acceptable.

File: src/test/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessageTest.java [hunk 7]
```diff
@@ -227,6 +228,23 @@ private static LocalizedMessage createSampleLocalizedMessage() {
...
 
+    @Test
+    public void testCompareToWithDifferentModuleId() {
+        final LocalizedMessage message1 = new LocalizedMessage(0,
+                "com.puppycrawl.tools.checkstyle.checks.coding.messages", "empty.statement",
+                EMPTY_OBJECT_ARRAY, "module1", LocalizedMessage.class, null);
+        final LocalizedMessage message2 = new LocalizedMessage(0,
+                "com.puppycrawl.tools.checkstyle.checks.coding.messages", "empty.statement",
+                EMPTY_OBJECT_ARRAY, null, LocalizedMessage.class, null);
+        final LocalizedMessage message3 = new LocalizedMessage(0,
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessageTest.java]: It's hard to follow what is the difference between these 3.
Rename `message3` to `message2` and rename current `message2` to `messageNull`.

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessageTest.java]: I'm also hesistant to say we should put the strings into local final variables to make reading slightly easier.

File: src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java (hunk 5 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java]: I saw every UT in this class used mock. I assumed there could be some reason so I just followed them.

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java (hunk 6 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java]: The super class uses `BriefUtLogger`. And it only print message like `/path/to/file.java message`. It don't have a module name or id postfix. We need smth like `/path/to/file.java message [ModuleName or ModuleId]`
So I change the logger and rewrite the verify logic. The verify code I wrote has some differences from the exist ones.

File: src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java (hunk 5 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java]: I don't know why this is, but since ever test is like this, we'll just leave it as is.

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java (hunk 6 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java]: @Luolc Can't we modify existing code in AbstractModuleTestSupport to supply an optional different Logger, one that prints full original message?

File: src/test/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessageTest.java (hunk 7 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessageTest.java]: Updated

File: src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java (hunk 3 above)
Comment (Luolc) [inline src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java (hunk 6 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java]: I was thinking that we only need this kind of verify here. If we don't ever need this anywhere else, is it necessary to modify the base class?
I could changing this into `AbstractModuleTestSupport`, but I preferred just write it in `CheckerTest`.

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java]: @romani I can hook in to most of the `verify` but there is some obstacles and weird things preventing a full hook because of the custom `AuditEventUtFormatter` instead of using the real `AuditEventDefaultFormatter`.

If you agree this should be fixed, lets move it to another issue.

**Comment by romani** [comment]
shippable is relaunched

**Comment by rnveach** [comment]
@romani Even though you already approved, please see my comments pinning you.

File: config/intellij-idea-inspections.xml (hunk 2 above)
Comment (romani) [inline config/intellij-idea-inspections.xml]: lets keep it for now, but will be addressed at #4929

File: src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java (hunk 4 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java]: moved to  #4930

File: src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java (hunk 5 above)
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AuditEventDefaultFormatterTest.java]: It was my mistake to allow such code to appear, mocks should not be used in simple cases like this.
new issue - #4931

File: src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java (hunk 6 above)
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java]: lets discuss this in separate issue .... over customization mean over complication .... all that is to please "test" code ... hmmm, lets discuss in detail what you propose , I am a bit suspicious.

**Comment by romani** [comment]
I commented each point.

**Comment by rnveach** [comment]
Thanks.
